### PR TITLE
Add authentication option to race list and add new API endpoints

### PR DESCRIPTION
**Pull request by Killklli**
Allows users to view unlisted races if they are authenticated

### CODE REVIEW

**Review by deains — COMMENTED** [review]
This is a bit of a roundabout way to list out the races. I don't think it's necessary to run two queries here, and it certainly isn't necessary to iterate over the unlisted races like you've done here. Instead you can do something like this:

```py
unlisted_filter = Q(unlisted=False)
if self.request.user.is_authenticated:
    unlisted_filter |= Q(unlisted=True, entrant__user=self.request.user)

races = models.Race.objects.filter(
    category__active=True,
).filter(unlisted_filter).exclude(state__in=[
    models.RaceStates.finished,
    models.RaceStates.cancelled,
])
```

File: racetime/templates/racetime/race_livesplit.html [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+<meta property="og:image" content="{{ meta_image }}"> {% endif %} {% endblock %} {% block header %}{% endblock header %} {% block nocurrentrace %}{% endblock nocurrentrace %} {% block footer %}{% endblock footer %} {% block main %}
+<style>
+    .timer {
+        display: none !important;
+    }
+    
+    .go-spectate {
+        display: none !important;
+    }
+</style>
```
Comment (deains) [inline racetime/templates/racetime/race_livesplit.html:24]: This section is a bit of a mess, could you clear it up a little? I think the `<meta>` tags are unnecessary here.

File: racetime/views/__init__.py [hunk 2]
```diff
@@ -179,11 +182,14 @@
...
     'RaceCSV',
     'RaceChatDelete',
+    'OAuthRaceChatDelete',
     'RaceChatLog',
     'RaceChatPurge',
+    'OAuthRaceChatPurge',
     'RaceData',
     'RaceListData',
     'RaceMini',
+    'RaceLivesplit',
```
Comment (deains) [inline racetime/views/__init__.py:192]: Pedantic, but it should really be `LiveSplit`.

File: racetime/views/race.py [hunk 3]
```diff
@@ -97,6 +97,10 @@ class RaceMini(Race):
     template_name_suffix = '_mini'
 
 
+class RaceLivesplit(Race):
+    template_name_suffix = '_livesplit'
```
Comment (deains) [inline racetime/views/race.py:101]: I think we should probably restrict this view to authenticated users only. I don't see any need for it if you're logged out, and it's not something I want Google to stumble on.

File: racetime/views/race.py [hunk 4]
```diff
@@ -160,6 +185,51 @@ def post(self, request, *args, **kwargs):
         return http.HttpResponse()
 
 
+@method_decorator(csrf_exempt, name='dispatch')
+class OAuthRaceChatDelete(ScopedProtectedResourceView, BotMixin, RaceChatMixin):
+    required_scopes = ['chat_delete']
```
Comment (deains) [inline racetime/views/race.py:190]: I think you should use the existing `race_action` scope for this.

File: racetime/views/race.py [hunk 5]
```diff
@@ -160,6 +185,51 @@ def post(self, request, *args, **kwargs):
...
+                'id': message.hashid,
+                'user': (
+                    message.user.api_dict_summary(race=race)
+                    if message.user else None
+                ),
+                'bot': message.bot.name if message.is_bot else None,
+                'is_bot': message.is_bot,
+                'deleted_by': self.user.api_dict_summary(race=race),
+            },
+        })
```
Comment (deains) [inline racetime/views/race.py:228]: There's obviously quite a bit of duplication here with `RaceChatDelete`, could you look into DRYing it up a little?

File: racetime/views/race.py [hunk 6]
```diff
@@ -200,6 +270,49 @@ def post(self, request, *args, **kwargs):
         return http.HttpResponse()
 
 
+@method_decorator(csrf_exempt, name='dispatch')
+class OAuthRaceChatPurge(ScopedProtectedResourceView, BotMixin, RaceChatMixin):
```
Comment (deains) [inline racetime/views/race.py:274]: Ditto the above remarks on this view.

**Review by deains — CHANGES_REQUESTED** [review]
Got some notes and thoughts below. 🙂 Very much looking forward to seeing the LiveSplit side of this!

**Comment by Killklli** [comment]
Also this deals with https://github.com/racetimeGG/racetime-app/issues/78 and https://github.com/racetimeGG/racetime-app/issues/48 and a handful of other issues not loaded into git

**Pull request merged**